### PR TITLE
Delay power until first window PoSt

### PR DESCRIPTION
**Pull request by Stebalien**
This will need a lot of tests, and the existing tests are now all broken.

# Design

Every partition gets a new `Unproven` bitfield for sectors that have yet to be proven and an `UnprovenPower` field store power to be proven and activated when the partition is next proven.

Crucially, new faulty power from unproven sectors is never removed from the miner, because it was never activated. Furthermore, unproven power is _added_ to the miner when it's proved the first time. To achieve this, many methods have gained `poweDelta` return values.

## Faults

When marked faulty, unproven sectors are removed from the unproven bitfield and added to the faulty bitfield. They now join the normal fault flow.

## Termination

Unproven sectors may be terminated (early) like any other sector.

## WindowPoSt

On WindowPoSt, skipped unproven sectors are added to the faulty sectors bitfield/power normally. Then, `UnprovenPower` and the `Unproven` bitfield are zeroed.

## End of deadline

At the end of the deadline, unproven sectors in unposted partitions are marked faulty along with their power.

Handling sector expiration before handling unproven sectors is explicitly forbidden.

### CODE REVIEW

File: actors/builtin/miner/deadline_state.go [hunk 1]
```diff
@@ -674,11 +685,14 @@ func (dl *Deadline) DeclareFaults(
...
 
-		newFaults, newPartitionFaultyPower, err := partition.DeclareFaults(store, sectors, sectorNos, faultExpirationEpoch, ssize, quant)
+		newFaults, partitionPowerDelta, partitionNewFaultyPower, err := partition.DeclareFaults(
+			store, sectors, sectorNos, faultExpirationEpoch, ssize, quant,
+		)
 		if err != nil {
 			return xerrors.Errorf("failed to declare faults in partition %d: %w", partIdx, err)
 		}
-		newFaultyPower = newFaultyPower.Add(newPartitionFaultyPower)
+		dl.FaultyPower = dl.FaultyPower.Sub(partitionNewFaultyPower)
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply actors/builtin/miner/deadline_state.go:699]: Good catch!

**Comment by Stebalien** [comment]
This still needs more tests, but it's now passing the existing tests and is ready to review.

File: actors/builtin/miner/deadline_state.go [hunk 2]
```diff
@@ -595,6 +597,15 @@ func (dl *Deadline) RemovePartitions(store adt.Store, toRemove bitfield.BitField
 			return xc.ErrIllegalArgument.Wrapf("cannot remove partition %d: has faults", partIdx)
 		}
 
+		// Don't allow removing partitions with unproven sectors.
+		hasNoUnproven, err := partition.Unproven.IsEmpty()
```
Comment (anorth) [inline actors/builtin/miner/deadline_state.go:601]: `allProven` to avoid the double negative below?

File: actors/builtin/miner/deadline_state_test.go [hunk 3]
```diff
@@ -54,11 +54,25 @@ func TestDeadlines(t *testing.T) {
...
 	addSectors := func(t *testing.T, store adt.Store, dl *miner.Deadline) {
-		power, err := dl.AddSectors(store, partitionSize, sectors, sectorSize, quantSpec)
+		activatedPower, err := dl.AddSectors(store, partitionSize, false, sectors, sectorSize, quantSpec)
 		require.NoError(t, err)
+		assert.True(t, activatedPower.IsZero())
+
+		sectorArr := sectorsArr(t, store, sectors)
+
+		// Prove everything
+		// TODO: test unproven power
```
Comment (anorth) [inline actors/builtin/miner/deadline_state_test.go:64]: Is this still TODO?

File: actors/builtin/miner/miner_state_test.go [hunk 4]
```diff
@@ -648,7 +648,8 @@ func TestSectorAssignment(t *testing.T) {
 
 		newPower, err := harness.s.AssignSectorsToDeadlines(harness.store, 0, sectorInfos, partitionSectors, sectorSize)
 		require.NoError(t, err)
-		require.True(t, newPower.Equals(miner.PowerForSectors(sectorSize, sectorInfos)))
+		require.True(t, newPower.IsZero())
+		//require.True(t, newPower.Equals(miner.PowerForSectors(sectorSize, sectorInfos)))
```
Comment (anorth) [inline actors/builtin/miner/miner_state_test.go:652]: Remove

File: actors/builtin/miner/miner_test.go [hunk 5]
```diff
@@ -2806,13 +2816,13 @@ func (h *actorHarness) confirmSectorProofsValid(rt *mock.Runtime, conf proveComm
...
-			}
-			rt.ExpectSend(builtin.StoragePowerActorAddr, builtin.MethodsPower.UpdateClaimedPower, &pcParams, big.Zero(), nil, exitcode.Ok)
-		}
+		//if !expectRawPower.IsZero() || !expectQAPower.IsZero() {
+		//	pcParams := power.UpdateClaimedPowerParams{
+		//		RawByteDelta:         expectRawPower,
+		//		QualityAdjustedDelta: expectQAPower,
+		//	}
+		//	rt.ExpectSend(builtin.StoragePowerActorAddr, builtin.MethodsPower.UpdateClaimedPower, &pcParams, big.Zero(), nil, exitcode.Ok)
+		//}
```
Comment (anorth) [inline actors/builtin/miner/miner_test.go:2825]: Remove

File: actors/builtin/miner/partition_state_test.go [hunk 6]
```diff
@@ -33,9 +33,12 @@ func TestPartitions(t *testing.T) {
 		store := ipld.NewADTStore(context.Background())
 		partition := emptyPartition(t, store)
 
-		power, err := partition.AddSectors(store, sectors, sectorSize, quantSpec)
+		// TODO: test unproven
```
Comment (anorth) [inline actors/builtin/miner/partition_state_test.go:36]: Done?

File: actors/builtin/miner/deadline_state_test.go (hunk 3 above)
Comment (Stebalien) [reply actors/builtin/miner/deadline_state_test.go:64]: Yes.

Comment (Stebalien) [reply actors/builtin/miner/deadline_state_test.go]: Ok, now it's done.

File: actors/builtin/miner/partition_state_test.go (hunk 6 above)
Comment (Stebalien) [reply actors/builtin/miner/partition_state_test.go:37]: Working on it.

**Comment by Stebalien** [comment]
Ok, I've found a bug. I'm not correctly handling power when handling missing posts at the end of the proving deadline. I'll upload a fix with tests tomorrow.

File: support/mock/mockrt.go [hunk 7]
```diff
@@ -328,10 +331,23 @@ func (rt *Runtime) Send(toAddr addr.Address, methodNum abi.MethodNum, params run
 	exp := rt.expectSends[0]
 
 	if !exp.Equal(toAddr, methodNum, params, value) {
```
Comment (Stebalien) [inline support/mock/mockrt.go:333]: Unrelated change to the testing framework to make reading errors easier.

**Comment by Stebalien** [comment]
@anorth the second commit fixes the bug (and adds/fixes more tests). I'm now explicitly returning:

* PowerDelta (not all newely faulty power is part of this delta, as some of it may not have been "active").
* PenaltyPower (retracted power + newly faulty power)
* NewFaultyPower

**Review by anorth — APPROVED** [review]
Thanks.

It's slightly unfortunate that the "policy" logic of which power to penalize is now in the state, but not a huge problem. We could probably resolve this and make things a bit more concise with a "power deltas" struct for use as return value from the methods that return more than one power thingy.

**Comment by Stebalien** [comment]
Hm. Yeah, adding a power delta object would (https://github.com/filecoin-project/specs-actors/issues/1011).
